tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php

```diff
@@ -0,0 +1,215 @@
+        self::assertEquals(
+            $expectedDiagnostics,
+            wait($provider->provideDiagnostics(new TextDocumentItem('file:///somefile.php', 'php', 1, $result->source())))
+        );
+    }
+
+    public function provideDiagnosticsTestData(): array
+    {
+        return [
+            'Empty file' => [
+                '<?php '
+            ],
+            'Class with no methods calls' => [
+                '<?php 
+				class Class1 
+				{
+					public function __construct() {
+					}
+
+					public function someOtherMethod() {
+						$var = 5;
+					}
+				}
+				'
+            ],
+            'Class with methods calls' => [
+                '<?php 
+				class Class1 
+				{
+					public function __construct() {
+					}
+
+					public function someOtherMethod() {
+						$var = 5;
+						$this->{{methodThatIsCalled}}()->{{otherMethod}}();
+						$this->{$var}();
+                        self::{{someStaticMethod}}();
+					}
+				}
+				'
+            ],
+        ];
+    }
+
+    public function testNoActionsAreProvidedWhenRangeIsNotAnInsertionPoint(): void
+    {
+        $provider = $this->createProvider();
+        self::assertEquals(
+            [],
+            wait($provider->provideActionsFor(new TextDocumentItem('file:///somefile.php', 'php', 1, '<?php $var = 5;'), new Range(
+                new Position(1, 7),
+                new Position(1, 9),
+            )))
+        );
+    }
+    /** @dataProvider provideActionsTestData */
+    public function testProvideActions(string $text, bool $shouldSucceed): void
+    {
+        $result = OffsetExtractor::create()
+            ->registerRange('diagnosticsRange', '{{', '}}')
+            ->registerOffset('selection', '<>')
+            ->parse($text);
+
+        $provider = $this->createProvider();
+        $selectionPosition = PositionConverter::byteOffsetToPosition($result->offset('selection'), $result->source());
+        $uri = 'file:///somefile.php';
+
+        if ($shouldSucceed) {
+            $expectedDiagnostic = new Diagnostic(
+                RangeConverter::toLspRange($result->range('diagnosticsRange'), $result->source()),
+                'Generate method',
+                DiagnosticSeverity::INFORMATION,
+                null,
+                'phpactor'
+            );
+            $codeActions = [
+                CodeAction::fromArray([
+                    'title' =>  'Generate method (if not exists)',
+                    'kind' => GenerateMethodProvider::KIND,
+                    'diagnostics' => [ $expectedDiagnostic ],
+                    'command' => new Command(
+                        'Generate method',
+                        GenerateMethodCommand::NAME,
+                        [
+                            $uri,
+                            $result->offset('selection')->toInt()
+                        ]
+                    )
+                ])
+            ];
+        } else {
+            $codeActions = [];
```

Comment:
please try and avoid `else` statements in the future. `if (!$shouldSucceed) { self::assertEmpty($blah) }`